code/code/misc/info.cc

```diff
@@ -5385,642 +5385,291 @@ void TBeing::describeTrapDamType(const TTrap *obj, int) const
-  TThing *primary=heldInPrimHand(), *secondary=heldInSecHand();
-  TThing *belt=equipment[WEAR_WAIST];
-  TThing *juju=equipment[WEAR_NECK];
-  TThing *wristpouch=equipment[WEAR_WRIST_R];
-  TThing *wristpouch2=equipment[WEAR_WRIST_L];
-  int totalcharges;
-  ritualismLevelT ritlevel = getRitualismLevel();
-
-  struct {
-    TThing *where;
-    ritualismLevelT ritlevel;
-  } search[] = {
-      {primary  , RIT_LEV_COMP_PRIM_OTHER_FREE},
-      {secondary, RIT_LEV_COMP_EITHER         },
-      {this    , RIT_LEV_COMP_INV            },
-      {belt     , RIT_LEV_COMP_BELT           },
-      {juju     , RIT_LEV_COMP_NECK           },
-      {wristpouch, RIT_LEV_COMP_WRIST         },
-      {wristpouch2, RIT_LEV_COMP_WRIST         }
+  auto isValidSpellType = [](skillUseClassT spellType) {
+    return std::find(validSpellTypes.begin(), validSpellTypes.end(), spellType) != validSpellTypes.end();
   };
 
-  if (!hasClass(CLASS_SHAMAN) && !isImmortal()) {
-    sendTo("You aren't a Shaman, therefore you have no use for rituals.\n\r");
-    sendTo("Perhaps using the SPELLS command is what you want.\n\r");
-    return;
-  }
+  static const std::map<skillUseClassT, unsigned short> classForSpellType = {
+    {SPELL_CLERIC, CLASS_CLERIC},
+    {SPELL_DEIKHAN, CLASS_DEIKHAN},
+    {SPELL_MAGE, CLASS_MAGE},
+    {SPELL_SHAMAN, CLASS_SHAMAN}
+  };
 
-  if (!(d = desc))
-    return;
+  auto countValidClasses = [this]() {
+    int count = 0;
+    for (auto spellType : validSpellTypes) {
+      if (hasClass(classForSpellType.at(spellType)))
+        count++;
+    }
+    return count;
+  };
 
-  *buffer = '\0';
+  std::map<skillUseClassT, std::map<spellCategories, std::vector<spellNumT>>> spellInfoByType;
+  for (auto spellType : validSpellTypes)
+    spellInfoByType.emplace(spellType, createSpellCategoriesMap());
 
-  if (argument.empty())
-    memset(types, 1, sizeof(int) * 4);      
-  else {
-    memset(types, 0, sizeof(int) * 4);
-
-    arg=argument.word(0);
-    arg2=argument.word(2);
-    arg3=argument.word(3);
-
-    if (is_abbrev(arg3, "all"))
-      showall = 1;
-
-    if (!arg2.empty()) {
-      if (is_abbrev(arg2, "all"))
-        showall = 1;
-      else if (is_abbrev(arg2, "targeted"))
-        subtype = 1;
-      else if (is_abbrev(arg2, "nontargeted"))
-        subtype = 2;
-      else
-        badtype = 1;
-    }
-    
-    if (is_abbrev(arg, "offensive")) {
-      if(!subtype || subtype == 1)
-        types[0] = 1;
-      if(!subtype || subtype == 2)
-        types[1] = 1;
-    } else if (is_abbrev(arg, "utility")) {
-      if(!subtype || subtype == 1)
-        types[2] = 1;
-      if(!subtype || subtype == 2)
-        types[3] = 1;      
-    } else
-      badtype = 1;
-    
-    if (badtype) {
-      sendTo("You must specify a valid ritual type.\n\r");
-      sendTo("Syntax: rituals <offensive|utility> <targeted|nontargeted> <all>.\n\r");
-      return;
-    }
-  }
+  std::vector<skillSorter> skillsSorted;
+  skillsSorted.reserve(230);
```

Comment:
We sure do. `discArray`, which gets iterated through to populate this vector, uses an array size of MAX_SKILL (452) + 1 so I could use that in the `reserve` call. Though that would definitely waste some memory, as it will probably never get close to that amount because many entries won't be added to the `skillsSorted` vector.

The other alternative would be for me to just break here in GDB and see how large the vector actually gets during runtime, then use that number + some padding (I believe that's where I got the number 230 in the first place, actually).

Reply:
So I checked by breaking here in GDB and the maximum value this vector gets to on an imm char with all skill/spells available is 234. Updated the value to 256 in https://github.com/sneezymud/sneezymud/pull/387/commits/2d8dd19e459348f1dd2aa71e74c6c0f4706b8edc just to be thorough.

Reply:
I'd be more worried about having a magic number in code than allocating 200 extra bytes of memory.

Reply:
Addressed in https://github.com/sneezymud/sneezymud/pull/387/commits/890af4a0b398e8f2776b35157714241057212f09